ATAC-seq preprocessing pipeline
  - Step 1: convert sra file to fastq file `sratools` `.fastq`
  - Step 2: Quality control `fastqc, fastp, cutadapt` `.fastq`
    - `fastqc` to check reads quality
    - `fastp` is a user friendly software which can remove adapator and low quality reads
    - `cutadapt` to remove adaptor
    - > what is 'adapator sequence'?
  - Step 3: Mapping `bowtie2` `.sam`
    - `bowtie2-build` create bowtie2 index with reference genome
    - `bowtie2` get result `.sam` of mapping
  - **Step 4**: Filter low-quality reads (MAPQ score < 10), then remove duplicate with `samtools` `.bam`
  - Step 5: Call peaks with `macs2` `.txt`
  - Step 6: Merge all peaks and generate gtf file `library(IReNA)` `.gtf`
  - Step 7: Calculate raw counts withn htseq `htseq-count` `.txt`
  - **Step 8**: Combine all samples, and calculate differential peaks according to edgeR `library(IReNA)` `differential_peaks.bed, peaks.bed`
  - Step 9: Merge bam files `samtools`
  - **Step 10**: Calculate footprints of merged bam files with HINT `rgt-hint` `differential_peaks.bed, filtered_footprints.bed`